File: verifiers/dotnet/Decionis.VerifyingProvider/VerifyingProvider.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace Decionis.VerifyingProvider;

/// <summary>
/// An independent .NET implementation of the Verifying Provider Profile
/// (docs/authority/verifying-provider.md in the agent-safe-pipeline repository),
/// VP-1 and VP-2, written against the profile's text and held to its vectors.
/// It shares no code with the executor; that is the point of a second
/// implementation.
/// </summary>
public sealed class VerifyingProvider
{
    public const string AttestationType = "decionis-claim-attestation+jwt";
    public const string JcsProfile = "RFC8785/JCS";
    private const string Label = "agentsafe";

    internal static readonly string[] BaseComponents =
    {
        "@method", "@path", "content-digest", "idempotency-key", "x-agent-safe-intent-hash",
    };

    internal static readonly string[] KnownComponents =
    {
        "@method", "@path", "content-digest", "idempotency-key", "x-agent-safe-intent-hash",
        "x-agent-safe-grant-id", "x-agent-safe-decision-id", "x-agent-safe-claim-attestation",
    };

    private readonly ProviderOptions options;

    public VerifyingProvider(ProviderOptions options)
    {
        this.options = options;
    }

    /// <summary>The RFC 9530 value the executor sends for a body, the empty body included.</summary>
    public static string ContentDigest(byte[]? body)
    {
        return "sha-256=:" + Convert.ToBase64String(SHA256.HashData(body ?? Array.Empty<byte>())) + ":";
    }

    /// <summary>The whole procedure for one received request.</summary>
    public Verdict Verify(ProviderRequest request)
    {
        var nowSeconds = options.Now().ToUnixTimeSeconds();
        if (!SignatureHolds(request, nowSeconds))
        {
            return Verdict.Refuse(Verdict.SignatureInvalidOrIncomplete);
        }
        if (!options.Effects)
        {
            return new Verdict(true, null, null);
        }
        ClaimAttestation attestation;
        try
        {
            attestation = AttestationOf(request.Headers.TryGetValue("x-agent-safe-claim-attestation", out var compact) ? compact : null);
        }
        catch (Exception e) when (e is FormatException or JsonException or ArgumentException or InvalidOperationException)
        {
            return Verdict.Refuse(Verdict.AttestationInvalid);
        }
        try
        {
            Described(attestation, request, nowSeconds);
        }
        catch (Exception e) when (e is ArgumentException or CanonicalDigest.NotIJsonException or InvalidOperationException or FormatException)
        {
            return Verdict.Refuse(Verdict.AttestationDoesNotDescribeThisRequest);
        }
        // The record is kept to the end of the second `exp` falls in, never shorter than `exp`.
        var expiresAt = DateTimeOffset.FromUnixTimeSeconds((long)Math.Ceiling(attestation.Exp));
        if (!options.Replay.Record(attestation.Sub, expiresAt))
        {
            return Verdict.Refuse(Verdict.GrantReplayed);
        }
        return new Verdict(true, null, attestation);
    }

    // Steps 0 to 5.
    private bool SignatureHolds(ProviderRequest request, long nowSeconds)
    {
        var headers = request.Headers;
        if (!headers.TryGetValue("signature-input", out var input) || !input.StartsWith(Label + "=", StringComparison.Ordinal))
        {
            return false;
        }
        var parameters = input[(Label.Length + 1)..];
        if (!headers.TryGetValue("signature", out var signatureHeader)
            || !signatureHeader.StartsWith(Label + "=:", StringComparison.Ordinal)
            || !signatureHeader.EndsWith(':'))
        {
            return false;
        }
        var encoded = signatureHeader[(Label.Length + 2)..^1];
        if (!headers.TryGetValue("content-digest", out var digest) || digest != ContentDigest(request.Body))
        {
            return false;
        }
        var covered = CoveredComponents(parameters);
        if (covered is null)
        {
            return false;
        }
        foreach (var component in options.Effects ? KnownComponents : BaseComponents)
        {
            if (!covered.Contains(component))
            {
                return false;
            }
        }
        var createdText = Parameter(parameters, ";created=");
        if (createdText is null || !long.TryParse(createdText, NumberStyles.None, CultureInfo.InvariantCulture, out var created))
        {
            return false;
        }
        var window = (long)options.ClockWindow.TotalSeconds;
        if (created < nowSeconds - window || created > nowSeconds + window)
        {
            return false;
        }
        var keyId = QuotedParameter(parameters, ";keyid=\"");
        var algorithm = QuotedParameter(parameters, ";alg=\"");
        ExecutorKey? key = null;
        foreach (var candidate in options.ExecutorKeys)
        {
            if (candidate.KeyId == keyId && candidate.Algorithm == algorithm)
            {
                key = candidate;
                break;
            }
        }
        if (key is null)
        {
            return false;
        }
        var baseBytes = SignatureBase(request, covered, parameters);
        if (baseBytes is null)
        {
            return false;
        }
        byte[] signature;
        try
        {
            signature = Convert.FromBase64String(encoded);
        }
        catch (FormatException)
        {
            return false;
        }
        switch (key.Algorithm)
        {
            case "ed25519":
                var verifier = new Ed25519Signer();
                verifier.Init(false, key.PublicKey);
                verifier.BlockUpdate(baseBytes, 0, baseBytes.Length);
                return signature.Length == 64 && verifier.VerifySignature(signature);
            case "hmac-sha256":
                var expected = HMACSHA256.HashData(key.Secret!, baseBytes);
                return CryptographicOperations.FixedTimeEquals(expected, signature);
            default:
                return false;
        }
    }

    /// <summary>The covered components a signature-input names, or null when the list is not one, names an unknown one, or one twice.</summary>
    internal static List<string>? CoveredComponents(string parameters)
    {
        if (!parameters.StartsWith('('))
        {
            return null;
        }
        var end = parameters.IndexOf(')', StringComparison.Ordinal);
        if (end < 0)
        {
            return null;
        }
        var inner = parameters[1..end];
        var names = new List<string>();
        if (inner.Length == 0)
        {
            return names;
        }
        foreach (var quoted in inner.Split(' '))
        {
            if (quoted.Length < 3 || quoted[0] != '"' || quoted[^1] != '"')
            {
                return null;
            }
            var name = quoted[1..^1];
            if (Array.IndexOf(KnownComponents, name) < 0 || names.Contains(name))
            {
                return null;
            }
            names.Add(name);
        }
        return names;
    }

    /// <summary>The base, exactly as both sides build it.</summary>
    internal static byte[]? SignatureBase(ProviderRequest request, List<string> covered, string parameters)
    {
        var lines = new List<string>();
        foreach (var component in covered)
        {
            string value;
            switch (component)
            {
                case "@method":
                    value = request.Method.ToUpperInvariant();
                    break;
                case "@path":
                    value = request.Path;
                    break;
                case "content-digest":
                    value = ContentDigest(request.Body);
                    break;
                default:
                    if (!request.Headers.TryGetValue(component, out var header))
                    {
                        return null;
                    }
                    value = header;
                    break;
            }
            lines.Add($"\"{component}\": {value}");
        }
        lines.Add($"\"@signature-params\": {parameters}");
        return Encoding.UTF8.GetBytes(string.Join("\n", lines));
    }

    private static string? Parameter(string parameters, string name)
    {
        var start = parameters.IndexOf(name, StringComparison.Ordinal);
        if (start < 0)
        {
            return null;
        }
        var rest = parameters[(start + name.Length)..];
        var end = rest.IndexOf(';', StringComparison.Ordinal);
        return end < 0 ? rest : rest[..end];
    }

    private static string? QuotedParameter(string parameters, string name)
    {
        var start = parameters.IndexOf(name, StringComparison.Ordinal);
        if (start < 0)
        {
            return null;
        }
        var rest = parameters[(start + name.Length)..];
        var end = rest.IndexOf('"', StringComparison.Ordinal);
        return end < 0 ? null : rest[..end];
    }

    // Step 6: the attestation by form, key, signature, issuer and the shape of
    // its claims; the claims when it is the authority's.
    private ClaimAttestation AttestationOf(string? compact)
    {
        var parts = (compact ?? "").Split('.');
        if (parts.Length != 3)
        {
            throw new ArgumentException("form");
        }
        using var header = JsonDocument.Parse(Base64Url.Decode(parts[0]));
        if (Text(header.RootElement, "alg") != "EdDSA" || Text(header.RootElement, "typ") != AttestationType)
        {
            throw new ArgumentException("header");
        }
        var key = options.AuthorityJwks.Find(Text(header.RootElement, "kid") ?? "") ?? throw new ArgumentException("kid");
        var signed = Encoding.ASCII.GetBytes(parts[0] + "." + parts[1]);
        var verifier = new Ed25519Signer();
        verifier.Init(false, key);
        verifier.BlockUpdate(signed, 0, signed.Length);
        var signature = Base64Url.Decode(parts[2]);
        if (signature.Length != 64 || !verifier.VerifySignature(signature))
        {
            throw new ArgumentException("signature");
        }
        using var document = JsonDocument.Parse(Base64Url.Decode(parts[1]));
        var claims = document.RootElement;
        if (claims.ValueKind != JsonValueKind.Object)
        {
            throw new ArgumentException("claims");
        }
        // The claims a provider compares, and the ones a receipt is built from,
        // must be there in the right type: an attestation without them is not one.
        if (!claims.TryGetProperty("binding", out var binding) || binding.ValueKind != JsonValueKind.Object)
        {
            throw new ArgumentException("binding");
        }
        if (!claims.TryGetProperty("exp", out var exp) || exp.ValueKind != JsonValueKind.Number)
        {
            throw new ArgumentException("exp");
        }
        var attestation = new ClaimAttestation(
            Required(claims, "iss"),
            Required(claims, "sub"),
            Required(claims, "decision_id"),
            Required(claims, "dossier_id"),
            Required(binding, "intent_hash"),
            Required(binding, "execution_payload_digest"),
            Required(binding, "execution_payload_canonicalization_profile"),
            Required(claims, "claim_token_digest"),
            Required(claims, "jti"),
            exp.GetDouble());
        if (attestation.Iss != options.AuthorityIssuer)
        {
            throw new ArgumentException("issuer");
        }
        return attestation;
    }

    // Step 7: whether the attestation describes this request.
    private static void Described(ClaimAttestation attestation, ProviderRequest request, long nowSeconds)
    {
        if (request.Body is null)
        {
            throw new ArgumentException("no body");
        }
        var digest = CanonicalDigest.Of(request.Body);
        var headers = request.Headers;
        if (!(headers.TryGetValue("x-agent-safe-grant-id", out var grant) && grant == attestation.Sub)
            || !(headers.TryGetValue("x-agent-safe-decision-id", out var decision) && decision == attestation.DecisionId)
            || !(headers.TryGetValue("x-agent-safe-intent-hash", out var intent) && intent == attestation.IntentHash)
            || attestation.ExecutionPayloadCanonicalizationProfile != JcsProfile
            || attestation.ExecutionPayloadDigest != digest
            || !(attestation.Exp > nowSeconds))
        {
            throw new ArgumentException("describes another request");
        }
    }

    private static string Required(JsonElement element, string name)
    {
        return Text(element, name) ?? throw new ArgumentException(name);
    }

    private static string? Text(JsonElement element, string name)
    {
        return element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;
    }
}

File: verifiers/spring/src/main/java/com/decionis/verifyingprovider/VerifyingProvider.java
package com.decionis.verifyingprovider;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.PublicKey;
import java.security.Signature;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * An independent implementation of the Verifying Provider Profile
 * (docs/authority/verifying-provider.md in the agent-safe-pipeline repository),
 * VP-1 and VP-2, written against the profile's text and held to its vectors.
 * It shares no code with the executor; that is the point of a second
 * implementation.
 */
public final class VerifyingProvider {
  public static final String ATTESTATION_TYPE = "decionis-claim-attestation+jwt";
  public static final String JCS_PROFILE = "RFC8785/JCS";
  private static final String LABEL = "agentsafe";

  static final List<String> BASE_COMPONENTS =
      List.of("@method", "@path", "content-digest", "idempotency-key", "x-agent-safe-intent-hash");
  static final List<String> GRANT_COMPONENTS =
      List.of("x-agent-safe-grant-id", "x-agent-safe-decision-id", "x-agent-safe-claim-attestation");
  static final List<String> KNOWN_COMPONENTS;

  static {
    List<String> known = new ArrayList<>(BASE_COMPONENTS);
    known.addAll(GRANT_COMPONENTS);
    KNOWN_COMPONENTS = List.copyOf(known);
  }

  private static final Pattern CREATED = Pattern.compile(";created=(\\d{1,12})(?:;|$)");
  private static final Pattern KEY_ID = Pattern.compile(";keyid=\"([^\"]*)\"(?:;|$)");
  private static final Pattern ALG = Pattern.compile(";alg=\"([a-z0-9-]+)\"(?:;|$)");
  private static final Pattern SIGNATURE = Pattern.compile("^agentsafe=:(.*):$");
  private static final ObjectMapper JSON = new ObjectMapper();

  private final ProviderOptions options;

  public VerifyingProvider(ProviderOptions options) {
    this.options = options;
  }

  /** The RFC 9530 value the executor sends for a body, the empty body included. */
  public static String contentDigest(byte[] body) {
    try {
      byte[] sum = MessageDigest.getInstance("SHA-256").digest(body == null ? new byte[0] : body);
      return "sha-256=:" + Base64.getEncoder().encodeToString(sum) + ":";
    } catch (Exception e) {
      throw new IllegalStateException(e);
    }
  }

  /** The whole procedure for one received request. */
  public Verdict verify(ProviderRequest request) {
    long nowSeconds = options.clock().instant().getEpochSecond();
    if (!signatureHolds(request, nowSeconds)) {
      return Verdict.refuse(Verdict.SIGNATURE_INVALID_OR_INCOMPLETE);
    }
    if (!options.effects()) {
      return new Verdict(true, null, null);
    }
    ClaimAttestation attestation;
    try {
      attestation = attestationOf(request.headers().get("x-agent-safe-claim-attestation"));
    } catch (Exception e) {
      return Verdict.refuse(Verdict.ATTESTATION_INVALID);
    }
    try {
      described(attestation, request, nowSeconds);
    } catch (Exception e) {
      return Verdict.refuse(Verdict.ATTESTATION_DOES_NOT_DESCRIBE_THIS_REQUEST);
    }
    // The record is kept to the end of the second `exp` falls in, never shorter than `exp`.
    Instant expiresAt = Instant.ofEpochSecond((long) Math.ceil(attestation.exp()));
    if (!options.replay().record(attestation.sub(), expiresAt)) {
      return Verdict.refuse(Verdict.GRANT_REPLAYED);
    }
    return new Verdict(true, null, attestation);
  }

  // Steps 0 to 5.
  private boolean signatureHolds(ProviderRequest request, long nowSeconds) {
    Map<String, String> headers = request.headers();
    String input = headers.get("signature-input");
    if (input == null || !input.startsWith(LABEL + "=")) {
      return false;
    }
    String parameters = input.substring(LABEL.length() + 1);
    Matcher encoded = SIGNATURE.matcher(headers.getOrDefault("signature", ""));
    if (!encoded.matches()) {
      return false;
    }
    if (!contentDigest(request.body()).equals(headers.get("content-digest"))) {
      return false;
    }
    List<String> covered = coveredComponents(parameters);
    if (covered == null) {
      return false;
    }
    for (String component : options.effects() ? KNOWN_COMPONENTS : BASE_COMPONENTS) {
      if (!covered.contains(component)) {
        return false;
      }
    }
    String createdText = group(CREATED, parameters);
    if (createdText == null) {
      return false;
    }
    long created = Long.parseLong(createdText);
    long window = options.clockWindow().getSeconds();
    if (created < nowSeconds - window || created > nowSeconds + window) {
      return false;
    }
    String keyId = group(KEY_ID, parameters);
    String algorithm = group(ALG, parameters);
    Optional<ExecutorKey> key = options.executorKeys().stream()
        .filter(candidate -> candidate.keyId().equals(keyId) && candidate.algorithm().equals(algorithm))
        .findFirst();
    if (key.isEmpty()) {
      return false;
    }
    byte[] base = signatureBase(request, covered, parameters);
    if (base == null) {
      return false;
    }
    byte[] signature;
    try {
      signature = Base64.getDecoder().decode(encoded.group(1));
    } catch (IllegalArgumentException e) {
      return false;
    }
    try {
      switch (key.get().algorithm()) {
        case "ed25519" -> {
          Signature verifier = Signature.getInstance("Ed25519");
          verifier.initVerify(key.get().publicKey());
          verifier.update(base);
          return verifier.verify(signature);
        }
        case "hmac-sha256" -> {
          Mac mac = Mac.getInstance("HmacSHA256");
          mac.init(new SecretKeySpec(key.get().secret(), "HmacSHA256"));
          return MessageDigest.isEqual(mac.doFinal(base), signature);
        }
        default -> {
          return false;
        }
      }
    } catch (Exception e) {
      return false;
    }
  }

  /** The covered components a signature-input names, or null when the list is not one, names an unknown one, or one twice. */
  static List<String> coveredComponents(String parameters) {
    if (!parameters.startsWith("(")) {
      return null;
    }
    int end = parameters.indexOf(')');
    if (end < 0) {
      return null;
    }
    String inner = parameters.substring(1, end);
    if (inner.isEmpty()) {
      return List.of();
    }
    List<String> names = new ArrayList<>();
    Set<String> seen = new HashSet<>();
    for (String quoted : inner.split(" ", -1)) {
      if (quoted.length() < 3 || quoted.charAt(0) != '"' || quoted.charAt(quoted.length() - 1) != '"') {
        return null;
      }
      String name = quoted.substring(1, quoted.length() - 1);
      if (!KNOWN_COMPONENTS.contains(name) || !seen.add(name)) {
        return null;
      }
      names.add(name);
    }
    return names;
  }

  /** The base, exactly as both sides build it. */
  static byte[] signatureBase(ProviderRequest request, List<String> covered, String parameters) {
    List<String> lines = new ArrayList<>();
    for (String component : covered) {
      String value;
      switch (component) {
        case "@method" -> value = request.method().toUpperCase(java.util.Locale.ROOT);
        case "@path" -> value = request.path();
        case "content-digest" -> value = contentDigest(request.body());
        default -> {
          value = request.headers().get(component);
          if (value == null) {
            return null;
          }
        }
      }
      lines.add("\"" + component + "\": " + value);
    }
    lines.add("\"@signature-params\": " + parameters);
    return String.join("\n", lines).getBytes(StandardCharsets.UTF_8);
  }

  // Step 6: the attestation by form, key, signature, issuer and the shape of its claims; the
  // claims when it is the authority's.
  private ClaimAttestation attestationOf(String compact) throws Exception {
    String[] parts = compact == null ? new String[0] : compact.split("\\.", -1);
    if (parts.length != 3) {
      throw new IllegalArgumentException("form");
    }
    Base64.Decoder base64url = Base64.getUrlDecoder();
    JsonNode header = JSON.readTree(base64url.decode(parts[0]));
    if (!"EdDSA".equals(header.path("alg").asText(null))
        || !ATTESTATION_TYPE.equals(header.path("typ").asText(null))) {
      throw new IllegalArgumentException("header");
    }
    PublicKey key = options.authorityJwks().find(header.path("kid").asText(""))
        .orElseThrow(() -> new IllegalArgumentException("kid"));
    Signature verifier = Signature.getInstance("Ed25519");
    verifier.initVerify(key);
    verifier.update((parts[0] + "." + parts[1]).getBytes(StandardCharsets.US_ASCII));
    if (!verifier.verify(base64url.decode(parts[2]))) {
      throw new IllegalArgumentException("signature");
    }
    JsonNode claims = JSON.readTree(base64url.decode(parts[1]));
    // The claims a provider compares, and the ones a receipt is built from, must be there in the
    // right type: an attestation without them is not one.
    JsonNode binding = claims.path("binding");
    JsonNode exp = claims.path("exp");
    if (!exp.isNumber()) {
      throw new IllegalArgumentException("exp");
    }
    ClaimAttestation attestation = new ClaimAttestation(
        text(claims, "iss"),
        text(claims, "sub"),
        text(claims, "decision_id"),
        text(claims, "dossier_id"),
        text(binding, "intent_hash"),
        text(binding, "execution_payload_digest"),
        text(binding, "execution_payload_canonicalization_profile"),
        text(claims, "claim_token_digest"),
        text(claims, "jti"),
        exp.asDouble());
    if (!options.authorityIssuer().equals(attestation.iss())) {
      throw new IllegalArgumentException("issuer");
    }
    return attestation;
  }

  // Step 7: whether the attestation describes this request.
  private static void described(ClaimAttestation attestation, ProviderRequest request, long nowSeconds)
      throws Exception {
    if (request.body() == null) {
      throw new IllegalArgumentException("no body");
    }
    String digest = CanonicalDigest.of(request.body());
    Map<String, String> headers = request.headers();
    if (!attestation.sub().equals(headers.get("x-agent-safe-grant-id"))
        || !attestation.decisionId().equals(headers.get("x-agent-safe-decision-id"))
        || !attestation.intentHash().equals(headers.get("x-agent-safe-intent-hash"))
        || !JCS_PROFILE.equals(attestation.executionPayloadCanonicalizationProfile())
        || !attestation.executionPayloadDigest().equals(digest)
        || !(attestation.exp() > nowSeconds)) {
      throw new IllegalArgumentException("describes another request");
    }
  }

  private static String text(JsonNode node, String field) {
    JsonNode value = node.path(field);
    if (!value.isTextual()) {
      throw new IllegalArgumentException(field);
    }
    return value.textValue();
  }

  private static String group(Pattern pattern, String text) {
    Matcher matcher = pattern.matcher(text);
    return matcher.find() ? matcher.group(1) : null;
  }
}
